File: commands/dominator.ts
import {
    SlashCommandBuilder,
    PermissionFlagsBits,
    ChatInputCommandInteraction,
} from "discord.js";
import { startSpan } from "@sentry/node";
import {
    ACTIONS,
    ATTRIBUTES,
    ATTR_PRETTY,
    ACTION_PRETTY,
    buildStringChoice,
    buildIntegerChoice,
} from "../clients/constants.js";
import {
    MessageDominators,
    MessageDominator,
} from "../clients/backend/dominator/messageDominators.js";
import {
    MemberDominators,
    MemberDominator,
} from "../clients/backend/dominator/memberDominators.js";
import embedDominator from "../embeds/dominator.js";

const data = new SlashCommandBuilder()
    .setName("dominator")
    .setDescription("Configure Sibyl's AutoMod Capabilities")
    .addSubcommand((subcommand) =>
        subcommand
            .setName("message")
            .setDescription("Configure Message Dominator")
            .addStringOption((option) =>
                option
                    .setName("attribute")
                    .setDescription("Attribute to Use")
                    .setRequired(true)
                    .addChoices(
                        ...ATTRIBUTES.map((attribute) =>
                            buildStringChoice(
                                ATTR_PRETTY[
                                    attribute as keyof typeof ATTR_PRETTY
                                ],
                                attribute
                            )
                        )
                    )
            )
            .addIntegerOption((option) =>
                option
                    .setName("action")
                    .setDescription("Action to Take Upon Exceeding Threshold")
                    .addChoices(
                        ...ACTIONS.map((action) =>
                            buildIntegerChoice(
                                ACTION_PRETTY[
                                    action as keyof typeof ACTION_PRETTY
                                ],
                                ACTIONS.indexOf(action)
                            )
                        )
                    )
            )
            .addNumberOption((option) =>
                option
                    .setName("threshold")
                    .setDescription("Threshold for Attribute Score")
                    .setMinValue(0)
                    .setMaxValue(1)
            )
    )
    .addSubcommand((subcommand) =>
        subcommand
            .setName("member")
            .setDescription("Configure Member Dominator")
            .addStringOption((option) =>
                option
                    .setName("attribute")
                    .setDescription("Attribute to Use")
                    .setRequired(true)
                    .addChoices(
                        ...ATTRIBUTES.concat([
                            "crime_coefficient_100",
                            "crime_coefficient_300",
                        ]).map((attribute) =>
                            buildStringChoice(
                                ATTR_PRETTY[
                                    attribute as keyof typeof ATTR_PRETTY
                                ],
                                attribute
                            )
                        )
                    )
            )
            .addIntegerOption((option) =>
                option
                    .setName("action")
                    .setDescription("Action to Take Upon Exceeding Threshold")
                    .addChoices(
                        ...ACTIONS.map((action) =>
                            buildIntegerChoice(
                                ACTION_PRETTY[
                                    action as keyof typeof ACTION_PRETTY
                                ],
                                ACTIONS.indexOf(action)
                            )
                        )
                    )
            )
            .addNumberOption((option) =>
                option
                    .setName("threshold")
                    .setDescription("Threshold for Attribute Score")
                    .setMinValue(0)
                    .setMaxValue(1)
            )
    );

async function execute(interaction: ChatInputCommandInteraction) {
    if (
        !interaction.memberPermissions?.has(PermissionFlagsBits.Administrator)
    ) {
        await interaction.reply({
            content:
                "You Do Not Have Permissions to Configure Notification Settings. You Must Have the Administrator Permission.",
            ephemeral: true,
        });
    } else {
        await interaction.deferReply();

        const type = interaction.options.getSubcommand();
        const dominator =
            type === "message" ? MessageDominators : MemberDominators;
        const action = interaction.options.getInteger("action");
        const threshold = interaction.options.getNumber("threshold");

        const attribute = interaction.options.getString("attribute")!;
        if (attribute === "crime_coefficient_100" && action) {
            await dominator.update({
                communityID: interaction.guildId,
                crime_coefficient_100_action: action,
            } as Partial<MemberDominator>);
        } else if (attribute === "crime_coefficient_300" && action) {
            await dominator.update({
                communityID: interaction.guildId,
                crime_coefficient_300_action: action,
            } as Partial<MemberDominator>);
        } else if (ATTRIBUTES.includes(attribute)) {
            const triggerData = { communityID: interaction.guildId } as
                | Partial<MessageDominator>
                | Partial<MemberDominator>;
            if (action !== null && action !== undefined)
                Object.assign(triggerData, { [`${attribute}_action`]: action });
            if (threshold !== null && threshold !== undefined)
                Object.assign(triggerData, {
                    [`${attribute}_threshold`]: threshold,
                });
            await dominator.update(triggerData);
        }

        interaction.editReply({
            embeds: [
                embedDominator(
                    (await dominator.read(interaction.guildId!)) as
                        | MessageDominator
                        | MemberDominator,
                    attribute,
                    interaction.client,
                    interaction.guild!
                ),
            ],
        });
        console.info(
            `${attribute.toUpperCase()} trigger has been successfully updated for ${type} Dominator in Channel: ${interaction.guild?.name} (${interaction.guildId})`
        );
    }
}

async function onExecute(interaction: ChatInputCommandInteraction) {
    startSpan(
        {
            name: "/dominator",
        },
        () => execute(interaction)
    );
}

export default { data, execute: onExecute };

File: commands/dominator.py
from typing import Union
from loguru import logger as log
from sentry_sdk import trace as sentry_trace
from discord import (
    ApplicationContext,
    Bot,
    Cog,
    SlashCommandGroup,
    option,
    SlashCommandOptionType,
    OptionChoice,
)
from clients.backend.dominator.member_dominators import MemberDominators
from clients.backend.dominator.message_dominators import MessageDominators
from clients.constants import ACTION, ATTR_PRETTY, ATTRIBUTE
from embeds.dominator import embed_dominator


class Dominator(Cog):

    base_dominator = SlashCommandGroup(
        "dominator", description="configure sibyl's automod capabilities"
    )

    def __init__(self, bot: Bot) -> None:
        self.bot = bot

    @base_dominator.command(description="configure message dominator")
    @sentry_trace
    @option(
        "attribute",
        type=SlashCommandOptionType.string,
        description="attribute to set",
        choices=list(
            map(
                lambda attr: OptionChoice(name=ATTR_PRETTY[attr.value], value=attr),
                ATTRIBUTE.__members__.values(),
            ),
        ),
    )
    @option(
        "action",
        type=ACTION,
        description="action to take upon exceeding threshold",
        required=False,
    )
    @option(
        "threshold",
        type=SlashCommandOptionType.number,
        description="threshold for attribute score",
        required=False,
        min_value=0,
        max_value=1,
    )
    async def message(
        self,
        ctx: ApplicationContext,
        attribute: str,
        action: ACTION,
        threshold: float,
    ) -> None:
        if not ctx.author.guild_permissions.administrator:
            await ctx.followup.send(
                "you do not have permissions to configure notification settings"
            )
            return
        await ctx.defer()

        await self.configure_dominator(
            ctx, MessageDominators, attribute, action, threshold
        )

    @base_dominator.command(description="configure member dominator")
    @sentry_trace
    @option(
        "attribute",
        type=SlashCommandOptionType.string,
        description="attribute to set",
        choices=list(
            map(
                lambda attr: OptionChoice(name=ATTR_PRETTY[attr], value=attr),
                ATTR_PRETTY.keys(),
            )
        ),
    )
    @option(
        "action",
        type=ACTION,
        description="action to take upon exceeding threshold",
        required=False,
    )
    @option(
        "threshold",
        type=SlashCommandOptionType.number,
        description="threshold for attribute score",
        required=False,
        min_value=0,
        max_value=1,
    )
    async def member(
        self,
        ctx: ApplicationContext,
        attribute: str,
        action: ACTION,
        threshold: float,
    ) -> None:
        if not ctx.author.guild_permissions.administrator:
            await ctx.followup.send(
                "you do not have permissions to configure notification settings"
            )
            return
        await ctx.defer()

        await self.configure_dominator(
            ctx, MemberDominators, attribute, action, threshold
        )

    async def configure_dominator(
        self,
        ctx: ApplicationContext,
        dominator: Union[MessageDominators, MemberDominators],
        attribute: str,
        action: ACTION,
        threshold: float,
    ) -> None:
        if attribute == "crime_coefficient_100" and action is not None:
            dominator.update(
                {"communityID": ctx.guild_id, "crime_coefficient_100_action": action}
            )
        elif attribute == "crime_coefficient_300" and action is not None:
            dominator.update(
                {"communityID": ctx.guild_id, "crime_coefficient_300_action": action}
            )
        else:
            trigger_data = {"communityID": ctx.guild_id}
            if action is not None:
                trigger_data[f"{attribute}_action"] = action
            if threshold is not None:
                trigger_data[f"{attribute}_threshold"] = threshold

            dominator.update(trigger_data)

        await ctx.edit(
            embed=embed_dominator(dominator.read(ctx.guild_id), attribute, ctx.guild)
        )

        log.info(
            "{} trigger has been successfully updated for {} in server: {} ({})",
            attribute,
            dominator.__class__.__name__,
            ctx.guild.name,
            ctx.guild_id,
        )
